STORM-2016 Topology submission improvement: support adding local jars and maven artifacts on submission (1.x)

HeartSaVioR opened the pull request:
* JIRA issue: http://issues.apache.org/jira/browse/STORM-2016
* design doc: https://cwiki.apache.org/confluence/display/STORM/A.+Design+doc%3A+adding+jars+and+maven+artifacts+at+submission
* discussion thread: http://mail-archives.apache.org/mod_mbox/storm-dev/201608.mbox/%3CCAF5108i9+tJaNZ0LgRkTMkVQEL7F+53k9uyzxcT6zhSU6OHx9Q@mail.gmail.com%3E

----

* bin/storm now supports "--jars" and "--packages" options
  * it's only effective with "storm jar" and "storm sql"
* introduce new module: storm-submit to help resolving dependencies with handling transitive dependencies
* StormSubmitter will upload dependencies to BlobStore when submitting topology
* Supervisor will download dependencies from BlobStore when such topology is assigned
* Supervisor will launch workers with adding downloaded dependencies to worker classpath

TODO

- [x] documentation
- [x] craft pull request against master branch

Btw, it might be better to place some modules out of external since 'external' has most of non storm-core modules and flux, sql, storm-submit, storm-kafka-monitor are not kind of connectors.

HeartSaVioR commented:
Changed to \[WIP] since there're failures of unit tests. I'll take care of that so please continue reviewing pull request.

HeartSaVioR commented:
Fixed build failure on supervisor_test.

HeartSaVioR commented:
Builds for PR against 1.x branch is failing and I fixed that from #1609. I'll rebase once #1609 is merged to 1.x-branch.

HeartSaVioR commented:
[STORM-2016](http://issues.apache.org/jira/browse/STORM-2016) can also resolve [STORM-1435](http://issues.apache.org/jira/browse/STORM-1435) via applying different strategy, specifying dependencies to --jars and --packages instead of creating uber jar.
(Yes in fact STORM-2016 is started to resolve STORM-1435.)

Note that `storm sql` adds storm-sql-core and its transitive dependencies to classpath to run Runner, and also adds storm-sql-runtime and its transitive dependencies to classpath to run topology by adding them to --jars option. 

With [STORM-2023](https://issues.apache.org/jira/browse/STORM-2023) (#1610), we can run the example of storm-sql by below command.

> ./bin/storm sql order_filtering.sql order_filtering --packages "org.apache.storm:storm-sql-kafka:1.1.1-SNAPSHOT,org.apache.storm:storm-kafka:1.1.1-SNAPSHOT,org.apache.kafka:kafka_2.10:0.8.2.2\!org.slf4j:slf4j-log4j12,org.apache.kafka:kafka-clients:0.8.2.2"

Without STORM-2023, we still have a chance to run the example of storm-sql by below command 

> ./bin/storm sql order_filtering.sql order_filtering --packages "org.apache.storm:storm-sql-kafka:1.1.1-SNAPSHOT,org.apache.storm:storm-kafka:1.1.1-SNAPSHOT,org.apache.kafka:kafka_2.10:0.8.2.2\!org.slf4j:slf4j-log4j12,org.apache.kafka:kafka-clients:0.8.2.2,org.apache.calcite:calcite-core:1.4.0-incubating\!commons-dbcp:commons-dbcp\!com.google.code.findbugs:jsr305\!org.codehaus.janino:janino\!org.codehaus.janino:commons-compiler\!org.pentaho:pentaho-aggdesigner-algorithm\!com.fasterxml.jackson.core:jackson-annotations"

revans2 commented on bin/storm.py:
```diff
@@ -154,6 +157,39 @@ def confvalue(name, extrapaths, daemon=True):
             return " ".join(tokens[1:])
     return ""
 
+def resolve_dependencies(packages):
```
nit: It would be nice to have a short circuit here so if there are no dependencies to resolve we don't have to call into another java application.

HeartSaVioR replied:
Good point. Will address.

HeartSaVioR replied:
Addressed.

revans2 commented:
Conceptually I like the idea, and overall the code looks good.  My biggest concern is around cleanup of the bobs after the topology finishes.  I don't see anywhere in the code that it is doing that, but I didn't dig too deeply into it.

satishd commented:
@revans2 This issue was raised in the proposal discussion [here](https://mail-archives.apache.org/mod_mbox/storm-dev/201608.mbox/%3CCAF5108hbsHL3MO6HU47ncayV%2BgcMZ1s-B2izvRFmL7ot-TcqJg%40mail.gmail.com%3E). I do not think that proposal handles removing the blobs.
@HeartSaVioR Do you want to address this issue in this PR or you have plans to address it later? 

I have not yet started reviewing this PR, I will try to review it soon.

HeartSaVioR commented:
@revans2 @satishd 
Yes I've excluded that for simplification. Since this is a new feature I'd love to show that this concept works.
Moreover, in order to do that Nimbus should be modified but I also concerned with Nimbus port work. If we (including @longdafeng) are OK to address this before porting I'll do that work in this PR.

Btw, since jars are not shared but packages and transitive dependencies are meant to be shared, we should handle them different. We should remove blobs which are passed to --jars.

abhishekagarwal87 commented:
@HeartSaVioR - not sure if it will be of help, but you can also checkout https://github.com/apache/storm/pull/1296/files. I was trying out something similar.

HeartSaVioR commented:
@abhishekagarwal87 
Thanks, I took a look and your modification is simpler.
Btw, I was thinking same approach, but there're some cases to handle while assembling jars (for example, the content of schema files for Spring modules should be all added into one file) so it doesn't work that cases. That's why I choose this approach. It's already used widely by Spark users indeed.

HeartSaVioR commented:
@revans2 @satishd 
Addressed removing jars when topology is killed. Please note that artifacts will be not removed since it's shared among whole topologies.
Since it modifies Nimbus, we can still merge this in without this feature, but I don't want to be blocked by the thing even pull request is not submitted.

HeartSaVioR commented:
Added documentation, and replace packages to artifacts since it's clearer for its meaning.

I think it's done with 1.x branch, so I'll go on porting this to master branch.

HeartSaVioR commented:
Porting this to master forces me to resolve huge conflict or modify the code manually. We shouldn't leave progress of Java port as it is.

manuzhang commented on bin/storm.py:
```diff
@@ -754,18 +827,24 @@ def parse_config_opts(args):
       CONFFILE = curr.pop()
+    elif token == "--jars":
+      jars_list.extend(curr.pop().split(','))
+    elif token == "--artifacts":
```
Is "artifacts" a maven specific word ? How about "dependencies" ?

HeartSaVioR replied:
No, the term `artifact` is also used from Ivy, too.
http://ant.apache.org/ivy/history/latest-milestone/terminology.html#artifact

I think term `dependencies` is closer to jars + artifacts since jars will be also added to classpath.

harshach replied:
whats the intention behind artifacts?

HeartSaVioR replied:
This is already documented for `jar` so you can refer `help jar` to see explanation. I've also documented to doc page.

The intention behind picking `artifacts` is that, I picked `packages` first since this is what Spark already used, but realized why they use `packages` as term is that they maintain spark-packages. So `artifacts` is the alternatives for me.
Btw, the term `artifact` is used by Maven and Ivy, and Aether also uses `Artifact` class to represent this.

manuzhang commented on bin/storm.py:
```diff
@@ -754,18 +827,24 @@ def parse_config_opts(args):
     elif token == "--config":
       global CONFFILE
       CONFFILE = curr.pop()
+    elif token == "--jars":
```
Are these two options common for commands ? If so, could we have some help info in the `print_command` function ?

HeartSaVioR replied:
It's only used for `storm jar` and `storm sql`, but showing them to `print_command` would be better. I'll address this.
Thanks for the suggestion.

HeartSaVioR replied:
It would be better to add its usage to comment doc of `jar` and `sql` since print_usage(jar) will print out the usage of jar, and so on sql.

manuzhang replied:
It's a bit strange to have `storm jar --jars`. Maybe `--dep-jars` ?

HeartSaVioR replied:
I think --jars is representing its meaning properly so I'd like to keep it as it is.
If there's other one which would want to change option to non-duplicated name, I'll do that.

manuzhang commented on docs/Command-line-client.md:
```diff
@@ -43,6 +43,19 @@ Syntax: `storm jar topology-jar-path class ...`
 
 Runs the main method of `class` with the specified arguments. The storm jars and configs in `~/.storm` are put on the classpath. The process is configured so that [StormSubmitter](javadocs/org/apache/storm/StormSubmitter.html) will upload the jar at `topology-jar-path` when the topology is submitted.
 
+When you want to ship other jars which is not included to application jar, you can pass them to `--jars` option with comma-separated string.
```
neither system provided nor included to application jar ?

manuzhang replied:
Could we have some examples here ?

HeartSaVioR replied:
That is not mandatory, but it only makes sense when that jar is not presented neither system provided nor included to application jar. If system or application jar provides that, that will be used. This is similar to precedence of classloading between system and application jar.

HeartSaVioR replied:
And yes we can have example on this. Will address.

manuzhang replied:
It's not clear whether system jar or application jar will get loaded and may create some surprise if the uploaded jar doesn't work.  Maybe some warnings somewhere.

HeartSaVioR replied:
It would be ideal if we can show same artifacts with different versions as warning, but I'm not sure it would be possible, especially application jar is an uber jar so we can't get artifacts within that jar.

If you mean we just would be better to show some general warnings, yes that would be good to have. Btw, this is closer to general behavior of determining worker classpath (even without these options), so it would be better to explain this in general manner.
I'm not sure where is the best place to address this. Let me think about it.

HeartSaVioR commented:
@manuzhang Thanks for checking the details of documentation. 
Please leave a comment when you're done reviewing of this. Thanks!

HeartSaVioR commented:
@revans2 @satishd @abhishekagarwal87 
Please review again when you get some time. I'll squash the commits and go on porting this to master branch once some of us give +1 for this.
Thanks in advance!

manuzhang commented on storm-core/src/jvm/org/apache/storm/generated/Assignment.java:
```diff
@@ -787,15 +787,15 @@ public void read(org.apache.thrift.protocol.TProtocol iprot, Assignment struct)
-                String _key549;
-                String _val550;
-                for (int _i551 = 0; _i551 < _map548.size; ++_i551)
+                org.apache.thrift.protocol.TMap _map564 = iprot.readMapBegin();
```
Forgive me for a dumb and irrelevant question. Why do we keep generated files in source ?

HeartSaVioR replied:
There's no dumb question. This is just a one of practices on Storm project. As always there're some pros and cons regarding this:

pros: Thrift compiler is only needed for contributors who modifies storm.thrift, and the case is actually not often.
cons: Thrift generates codes which makes huge diff even though just adding only one field.

IMO, installing thrift compiler because of just build project is bad, especially OSPs are using different versions of thrift. While disadvantage is also not a trivial, I think advantage is valuable enough.

manuzhang commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -893,29 +891,61 @@
+    (every? #(Utils/checkFileExists target-dir %) file-names)
+    true))
+
+(defn download-blobs-in-blobstore-map-for-topology-succeed?
```
It seems these methods could be private ?

HeartSaVioR replied:
I don't mind if it's private or not, since we already ported this to Java in master branch.
There're other functions which could be private.

manuzhang commented on storm-core/src/clj/org/apache/storm/daemon/supervisor.clj:
```diff
@@ -875,14 +875,12 @@
-  [conf stormconf-path localizer tmproot]
-  (let [storm-conf (read-supervisor-storm-conf-given-path conf stormconf-path)
-        blobstore-map (storm-conf TOPOLOGY-BLOBSTORE-MAP)
+  [conf storm-conf localizer tmproot blobs fn-local-resources fn-symlink-name]
```
update the comments instead of removing it ?

HeartSaVioR replied:
OK I'll do.

manuzhang commented on storm-core/src/clj/org/apache/storm/config.clj:
```diff
@@ -141,6 +141,10 @@
   [conf stormconf-path]
   (merge conf (clojurify-structure (Utils/fromCompressedJsonConf (FileUtils/readFileToByteArray (File. stormconf-path))))))
 
+(defn read-supervisor-storm-code-given-path
```
maybe update the `read-supervisor-topology` function below to use this

HeartSaVioR replied:
Great point! Will address.

manuzhang commented on external/storm-submit/src/main/java/org/apache/storm/submit/command/DependencyResolverMain.java:
```diff
@@ -0,0 +1,105 @@
+
+    public static void main(String[] args) {
+        if (args.length < 1) {
+            throw new IllegalArgumentException("packages must be presented.");
```
artifacts now ? Same for error message below

HeartSaVioR replied:
Great finding. Will address.

manuzhang commented on external/storm-submit/src/main/java/org/apache/storm/submit/dependency/AetherUtils.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public static Dependency parseDependency(String dependency) {
+        List<String> dependencyAndExclusions = Arrays.asList(dependency.split("!"));
```
"!" is a special bach character that refers to previous command. Although we can escape it with "\", is there a better choice ?

HeartSaVioR replied:
'!' occasionally represents 'not', so I think it would be easier to be recognized as excluded. Do we feel similar from '~'?

Refer to http://www.gnu.org/software/bash/manual/bashref.html#Quoting we should avoid to use '$', '`', '\', '"', '!', '@', '*' in double quotes if we want to avoid escape.

manuzhang replied:
another common character for **not** is `^`

HeartSaVioR replied:
Personally I prefer `!`, but if escaping matters I think `^` is also fine. I'll address this.

HeartSaVioR replied:
Addressed.

manuzhang commented on external/storm-submit/src/main/java/org/apache/storm/submit/dependency/AetherUtils.java:
```diff
@@ -0,0 +1,81 @@
+        buffer.append(artifact.getGroupId());
+        buffer.append(':').append(artifact.getArtifactId());
+        buffer.append(':').append(artifact.getExtension());
+        if(artifact.getClassifier().length() > 0) {
```
nit: space here

HeartSaVioR replied:
OK will address.

satishd commented on storm-core/src/jvm/org/apache/storm/dependency/DependencyUploader.java:
```diff
@@ -0,0 +1,166 @@
+            // as a workaround, we call getBlobMeta() for all keys
+            blobStore.getBlobMeta(key);
+        } catch (KeyNotFoundException e) {
+            // TODO: do we want to add ACL here?
```
Can we file JIRA for this? This is a low priority to have any acls user may want to provide.

HeartSaVioR replied:
Before that, we need to check that nimbus and supervisor can access these blobs. I didn't look deeply with blobstore so I would like someone to confirm this.
@revans2 Could you guide this thing?

satishd replied:
I do not think we should have acls for artifacts. If it is possible to support acls in this scenario, it should be done only for jars option.

satishd commented on storm-core/src/jvm/org/apache/storm/StormSubmitter.java:
```diff
@@ -233,6 +236,23 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
+                try {
+                    uploader.init();
+                    jarsBlobKeys = uploadDependencyJarsToBlobStore(uploader);
+                    artifactsBlobKeys = uploadDependencyArtifactsToBlobStore(uploader);
```
Both `jarsBlobKeys` and `artifactsBlobKeys` should be removed from blob store when topology submission is failed.

HeartSaVioR replied:
We can do that for jarsBlobKeys but can't do that for artifactsBlobKeys since we share artifact blobs and we skip uploads if there're same artifact uploaded before.

I'll address this for jarsBlobKeys. Thanks!

satishd commented:
@HeartSaVioR Overall LGTM. I will revisit once the comments are addressed/resolved.

HeartSaVioR commented:
@manuzhang Thanks, I addressed your latest comments from https://github.com/apache/storm/pull/1608/commits/a9f4886fd2f093c7a4450b55ef809fc9826c820e

Not sure we have better representation than '!' without conflicting bash special char and also not used char for maven artifact.

@satishd Thanks, I addressed your last comment from https://github.com/apache/storm/pull/1608/commits/17cb2dd333205578bed8f8af8a51072b18daf32d

For custom ACL for jars blobs I'd rather see possible issues first when we set ACL to blobs. If it's no issue at all and we agree that there could be a case of this I'll file an issue.
Please take a look again. Thanks!

satishd commented on storm-core/src/jvm/org/apache/storm/StormSubmitter.java:
```diff
@@ -238,39 +239,24 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
+                    submitTopologyInDistributeMode(name, topology, opts, progressListener, asUser, conf, serConf);
+                } catch (Throwable e) {
+                    // remove uploaded jars blobs, not artifacts since they're shared across the cluster
+                    uploader.deleteBlobs(jarsBlobKeys);
```
InvalidTopologyException and AlreadyAliveException definitely indicates that the topology is not successfully deployed and respective jars can be removed from blob store.
 
@HeartSaVioR 
But this code assumes that any Exception thrown here indicates that the submitted topolgy is not deployed successfully.  I am not sure whether that assumption is always right.

HeartSaVioR replied:
@satishd 
In fact, submitTopologyWithOpts() can also throw AuthorizationException and org.apache.thrift.TException, too.

Let's back to flow of logic.
Please note that we didn't catch Throwable (some known Exceptions and also RuntimeException) and just pass to caller. And there's also no return for Nimbus.Client.submitTopology(), which indicates that unless we get TException or other Exceptions it should be treated as succeed. (If not we can't guarantee topology deploy is succeed.)

I think this assumption is right for now, and assumption should be right unless we add some operations after calling submitTopologyInDistributeMode() or adding some operations after logging "Finished submitting topology".

satishd replied:
@HeartSaVioR Is not there a possibility that the topology is submitted successfully on nimbus but client receives thrift error because of some timeout or connection error? 
My preference is to remove those topologies only when we know definitely that the topology submission is failed.

HeartSaVioR replied:
@satishd 
Yes I think the scenarios you stated can be occurred. But also Nimbus throws TException when Nimbus is unable to handle the submission, say, runtime exception occurred.

So it's more like loosely catching with less risk vs hardly catching with more risk.
(Risk could be reduced via checking topology name once again indeed.)

After thinking it again, leaving blobs seems less harmful than whole contents of topology couldn't be downloaded at all. I'll follow your suggestion.

HeartSaVioR replied:
Addressed.

satishd commented:
+1
@HeartSaVioR Nice work.

HeartSaVioR commented:
@satishd 
Thanks for the thoughtful review and vote. I rebased the commits into one.

@revans2 @manuzhang @abhishekagarwal87 
Please continue reviewing. I'll merge this by tomorrow if there's no more discussion. Thanks.

harshach commented:
@HeartSaVioR can you give it 2 days given that its a big patch. Thanks.

HeartSaVioR commented:
@harshach Sure, I'll wait for the review during this week. Btw, Actual modification is around (or less than) 1000 lines, not that much.

manuzhang commented:
@HeartSaVioR The code looks good (non-binding). I tried running `storm sql` but ended up with timeout exceptions 

```
Exception in thread "main" java.lang.RuntimeException: org.sonatype.aether.resolution.DependencyResolutionException: Failed to collect dependencies for org.apache.storm:storm-sql-kafka:jar:1.1.1-SNAPSHOT (compile)
        at org.apache.storm.submit.command.DependencyResolverMain.main(DependencyResolverMain.java:63)
Caused by: org.sonatype.aether.resolution.DependencyResolutionException: Failed to collect dependencies for org.apache.storm:storm-sql-kafka:jar:1.1.1-SNAPSHOT (compile)
        at org.sonatype.aether.impl.internal.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:371)
        at org.apache.storm.submit.dependency.DependencyResolver.resolve(DependencyResolver.java:67)
        at org.apache.storm.submit.command.DependencyResolverMain.main(DependencyResolverMain.java:52)
Caused by: org.sonatype.aether.collection.DependencyCollectionException: Failed to collect dependencies for org.apache.storm:storm-sql-kafka:jar:1.1.1-SNAPSHOT (compile)
        at org.sonatype.aether.impl.internal.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:258)
        at org.sonatype.aether.impl.internal.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:333)
        ... 2 more
Caused by: org.sonatype.aether.resolution.ArtifactDescriptorException: Failed to read artifact descriptor for org.apache.kafka:kafka_2.10:jar:0.8.2.2\
        at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:282)
        at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor(DefaultArtifactDescriptorReader.java:172)
        at org.sonatype.aether.impl.internal.DefaultDependencyCollector.process(DefaultDependencyCollector.java:412)
        at org.sonatype.aether.impl.internal.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:240)
        ... 3 more
Caused by: org.sonatype.aether.resolution.ArtifactResolutionException: Could not transfer artifact org.apache.kafka:kafka_2.10:pom:0.8.2.2\ from/to central (http://repo1.maven.org/maven2/): Error transferring file: Connection timed out
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:537)
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:216)
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:193)
        at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:267)
        ... 6 more
Caused by: org.sonatype.aether.transfer.ArtifactTransferException: Could not transfer artifact org.apache.kafka:kafka_2.10:pom:0.8.2.2\ from/to central (http://repo1.maven.org/maven2/): Error transferring file: Connection timed out
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$4.wrap(WagonRepositoryConnector.java:975)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$4.wrap(WagonRepositoryConnector.java:966)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$GetTask.flush(WagonRepositoryConnector.java:707)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$GetTask.flush(WagonRepositoryConnector.java:701)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector.get(WagonRepositoryConnector.java:452)
        at org.sonatype.aether.impl.internal.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:456)
        ... 9 more
Caused by: org.apache.maven.wagon.TransferFailedException: Error transferring file: Connection timed out
        at org.apache.maven.wagon.providers.http.LightweightHttpWagon.fillInputData(LightweightHttpWagon.java:143)
        at org.apache.maven.wagon.StreamWagon.getInputStream(StreamWagon.java:116)
        at org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:88)
        at org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61)
        at org.sonatype.aether.connector.wagon.WagonRepositoryConnector$GetTask.run(WagonRepositoryConnector.java:615)
        at org.sonatype.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:60)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection timed out
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at java.net.Socket.connect(Socket.java:538)
        at sun.net.NetworkClient.doConnect(NetworkClient.java:180)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
        at sun.net.www.http.HttpClient.<init>(HttpClient.java:211)
        at sun.net.www.http.HttpClient.New(HttpClient.java:308)
        at sun.net.www.http.HttpClient.New(HttpClient.java:326)
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1169)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1105)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:999)
        at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:933)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1513)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1441)
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
        at org.apache.maven.wagon.providers.http.LightweightHttpWagon.fillInputData(LightweightHttpWagon.java:115)
        ... 8 more
```
I may have some problems connecting to maven central but why the resolver didn't look at my local repo which did contain the required dependencies.

HeartSaVioR commented:
@manuzhang 1.1.1-SNAPSHOT is not officially released (we don't release SNAPSHOT officially) so you need to build entire Storm code with `mvn clean install` before doing that. We can change the example to 1.1.0 or so after the feature is released to official.

manuzhang commented:
@HeartSaVioR yes, I did so and verified I have all the dependencies locally. I also modified the `DependencyResolverTest` to resolve `1.1.1-SNAPSHOT` and the UT passed.

HeartSaVioR commented:
@manuzhang OK great. Thanks for the update.

HeartSaVioR commented:
@manuzhang A bit confusing so would like to double check on this. Do you succeed to resolve 1.1.1-SNAPSHOT after `mvn clean install`? Or you're still failing?

manuzhang commented:
The exception is caused by the escaping "\" character previously. Things are working now after I updated to the latest version.

harshach commented on bin/storm.py:
```diff
@@ -232,44 +271,91 @@ def jar(jarfile, klass, *args):
+    local_jars = DEP_JARS_OPTS
+    artifact_to_file_jars = resolve_dependencies(DEP_ARTIFACTS_OPTS)
+
+    sql_core_jars = get_jars_full(STORM_DIR + "/external/sql/storm-sql-core")
```
making this hardcoded path to external code doesn't sound right.  we should make sql as top-level directory

HeartSaVioR replied:
Definitely. I've also stated that thing to somewhere other pull requests.
`external` has various kinds of modules, and we need to arrange that. 

Maybe we can leave kind of connectors as it is (we can even rename on this), and get flux, sql, kafka-monitor (this module is referred from script in bin, too), and more to out of external.

Since this issue is not only regarding to storm-sql I would like to initiate the discussion before making the change, and handle this to another issue. What do you think?

harshach replied:
@HeartSaVioR sounds good to me. We can address this in another JIRA.

harshach commented:
+1

HeartSaVioR commented:
Please hold off merging. I'll create pull request against master. We can merge both of them together.

HeartSaVioR commented:
OK, I created pull request against master: #1626

HeartSaVioR commented:
@harshach @satishd @manuzhang 
I just fixed one thing from nimbus.clj: just log exception instead of crashing nimbus when `blob-rm-dependency-jars-in-topology` throws Exception. Before that nimbus test is failing intermittent.

You can just review nimbus.clj once again. Thanks in advance!

HeartSaVioR commented:
I found a bug while testing other stuff with applied cluster. 
StormSubmitter doesn't handle empty --jars option properly. Will address this shortly both this and PR for master branch.

HeartSaVioR commented:
Done. Fixed things after receiving +1 are here:

1. nimbus.clj: just log exception instead of crashing nimbus when blob-rm-dependency-jars-in-topology
2. DependencyPropertiesParser.java: check --jars parameter string is empty, and treat it as empty list (also add new unit test testing this change)

HeartSaVioR commented:
Rebased with current 1.x-branch.

HeartSaVioR commented:
Rebased with current 1.x-branch again.

HeartSaVioR commented:
Changed module name as suggestion from pull request on master branch.

HeartSaVioR commented:
Merged via df8346ce1a61ccad2a0c3f715b6c3a53bdaf5670
